cellarium/ml/models/onepass_mean_var_std.py

```diff
@@ -64,9 +70,28 @@ def forward(self, x_ng: torch.Tensor, var_names_g: np.ndarray) -> dict[str, torc
         assert_columns_and_array_lengths_equal("x_ng", x_ng, "var_names_g", var_names_g)
         assert_arrays_equal("var_names_g", var_names_g, "var_names_g", self.var_names_g)
 
-        self.x_sums = self.x_sums + x_ng.sum(dim=0)
-        self.x_squared_sums = self.x_squared_sums + (x_ng**2).sum(dim=0)
-        self.x_size = self.x_size + x_ng.shape[0]
+        if self.algorithm == "naive":
+            self.x_sums = self.x_sums + x_ng.sum(dim=0)
+            self.x_squared_sums = self.x_squared_sums + (x_ng**2).sum(dim=0)
+            self.x_size = self.x_size + x_ng.shape[0]
+        elif self.algorithm == "shifted_data":
+            if self.x_shift is None:
```

Comment:
So... you get shift from the very first minibatch? I think it is an improvement over naive, but doesn't really close the issue ;-) we should implement Welford's online algorithm for that: https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance

Reply:
I have worked on that initially but then it turned out to be that it is either unstable or hard to implement merging results from multiple GPUs. Can't remember which one it was 🙃